add LSP capability textDocument/documentHighlight

lgo opened the pull request:
@jvilk-stripe -- I've submitted the PR to be able to pull the CI release build. I haven't tries the LSP functionality yet but a review pass would be great since the LSP tests seem ✅ locally.

This adds the LSP functionality for `textDocument/documentHighlight` ([Microsoft specification for documentHighlight](https://microsoft.github.io/language-server-protocol/specification#textDocument_documentHighlight)).

When a user selects a piece of text, the LSP responds with a list of ranges to highlight in the current file which relate to the initiate selection. 

Currently, highlighting uses the same information `textDocument/references` does. That means  this will work for typed files otherwise it can only handle Ruby constants in untyped files. Compared to regular fuzzy-highlighting, this will only respond with highlights for an identifier reference within the scope it is valid in which is the added benefit!

This does not implement the response field `kind?: DocumentHighlightKind`, which identifies the response values as the more specific highlights:
- `Text`, a textual occurrence -- typically a fuzzy response.
- `Read`, a read-access of a symbol, like reading a variable.
- `Write`, a write-access of a symbol, like writing to a variable.

By default, the editor will consider no `kind` as a `Text` response.

After poking in VSCode I noticed there are [distinguishing styling for the types](https://github.com/microsoft/vscode/blob/466e5ef6c2188d61a8c93c53748aa9e9788bcfc9/src/vs/editor/contrib/wordHighlighter/wordHighlighter.ts#L427-L452). In particular, the styling settings, [found here](https://code.visualstudio.com/api/references/theme-color#editor-colors), are:
- `editor.wordHighlight{Background,Border}` for read-access
- `editor.wordHighlightStrong{Background,Border}` for write-access
- `editor.selectionHighlight{Background,Border}` for fuzzy text.

It wasn't obvious whether that information could be pulled out of the LSP query `DispatchResult` so I didn't dig any further.

### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->
LSP features :)


### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->
Updated the tests so that all of the `def` + `usage` assertions will also check the highlight query. It differs by only checking the query returns values for all assertions in one file rather than all of them.

DarkDimius commented on main/lsp/requests/document_highlight.cc:
```diff
@@ -0,0 +1,91 @@
+        response->result = variant<JSONNullObject, vector<unique_ptr<DocumentHighlight>>>(JSONNullObject());
+        auto &queryResponses = result.responses;
+        if (!queryResponses.empty()) {
+            const bool fileIsTyped = uri2FileRef(uri).data(*gs).strictLevel >= core::StrictLevel::True;
```
this will fail enforce inside `data` if `uri` is non-existing.
Le'ts check for 
```
auto file = uri2FileRef(uri)
if (!file.exists())  {
  // handle 
}
```

DarkDimius commented on main/lsp/lsp.cc:
```diff
@@ -57,24 +57,30 @@ LSPLoop::QueryRun LSPLoop::setupLSPQueryByLoc(unique_ptr<core::GlobalState> gs,
 
-LSPLoop::QueryRun LSPLoop::setupLSPQueryBySymbol(unique_ptr<core::GlobalState> gs, core::SymbolRef sym) const {
+LSPLoop::QueryRun LSPLoop::setupLSPQueryBySymbol(unique_ptr<core::GlobalState> gs, core::SymbolRef sym,
+                                                 optional<string_view> uri) const {
```
let's give `uri` a more descriptive name? `overSingleFile`?

DarkDimius commented:
> I've submitted the PR to be able to pull the CI release build

I kicked off build. Note that it by default doesn't build Mac version for branches, so, I guess you're using a mac?

You should be able to build it if you want by following https://github.com/sorbet/sorbet#building-sorbet

```
bazel build //main:sorbet --config=release-mac
```

note that for practical purposes, I'd suggest to build

```
bazel build //main:sorbet --config=dbg --config=static-libs
```
instead, as it will allow you to easily debug if something goes wrong(though will be substantially slower)

DarkDimius commented on main/lsp/lsp.cc:
```diff
@@ -57,24 +57,30 @@ LSPLoop::QueryRun LSPLoop::setupLSPQueryByLoc(unique_ptr<core::GlobalState> gs,
-             std::find(usedConstants.begin(), usedConstants.end(), symNameHash) != usedConstants.end())) {
-            frefs.emplace_back(ref);
+    if (uri.has_value()) {
+        auto fref = uri2FileRef(uri.value());
```
ditto about `uri2FileRef`, let's handle the case when `uri` has invalid value.

DarkDimius commented on main/lsp/requests/document_highlight.cc:
```diff
@@ -0,0 +1,91 @@
+            // N.B.: Ignores literals.
+            // If file is untyped, only supports find reference requests from constants and class definitions.
+            if (auto constResp = resp->isConstant()) {
+                std::vector<std::unique_ptr<Location>> locations;
```
this is unused and thus will error the build(we error on unused variables)

lgo replied:
Is there something different from the CI builds? I was using `--config=release-mac` and it seemed to happily build with this unused variable.

DarkDimius commented on main/lsp/requests/document_highlight.cc:
```diff
@@ -0,0 +1,91 @@
+                tie(gs, response->result) = getHighlightsToSymbolInFile(move(gs), uri, constResp->symbol);
+            } else if (auto defResp = resp->isDefinition()) {
+                if (fileIsTyped || defResp->symbol.data(*gs)->isClass()) {
+                    std::vector<std::unique_ptr<Location>> locations;
```
ditto

DarkDimius commented:
> It wasn't obvious whether that information could be pulled out of the LSP query DispatchResult so I didn't dig any further.


Even if it can't currently, it can be added, but let's not do it in MVP :-)

lgo commented:
Addressed the comments. Also, I put the feature behind an experimental flag,

I wanted the build for an easy way to get the LSP set up for pay-server since I wasn't able to get it starting, but the LSP worked fine for a single file project 🎉.

With Sorbet disabled
![withoutDocumentHighlight](https://user-images.githubusercontent.com/1161183/63230754-f1be2800-c1c5-11e9-8c46-78849253a2d1.gif)

With Sorbet enabled and the experimental flag for highlight.
![withDocumentHighlight](https://user-images.githubusercontent.com/1161183/63230758-f8e53600-c1c5-11e9-86fd-71e175436d9e.gif)

Method highlight, which is a little weird! It highlights the whole `def ...` line.
![methodDocumentHighlight](https://user-images.githubusercontent.com/1161183/63230761-07335200-c1c6-11e9-9166-a7ebb201c19f.gif)

I'm going to see what behaviour LSPs for Elixir or Crystal have, if they implement this, and see how it feels.

DarkDimius commented:
> I'm going to see what behaviour LSPs for Elixir or Crystal have

I'd also suggest looking at TypeScript, they have probably the best LSP integration out there. I haven't heard particularly good things about these two.

lgo commented:
Good point! I was thinking in terms of syntactical similarity. This turned out to be quite easy to verify what was going on since VSCode has a setting to verbosely log the LSP messages`"typescript.tsserver.trace": "verbose"`. :)

Here's what I found: https://gist.github.com/lgo/e93e690d67db3ceed4d5701e38139251

The current quirks in the sorbet implementation are
- highlighting entire method definition lines (`def` + args)
- highlighting entire method call line (receiver, inputs, input block), eg: `d.method(1, 2) {}`. the same happens for `send` assignments assignments, eg: `d.method = 'bar'`)
- highlighting the `:` when highlighting parameter identifiers, eg: `foo(b:)`

And some other nice-to-haves would probably be highlighting exact string literals and symbol literals.

This example `rb` file highlights those quirks: https://gist.github.com/lgo/596ea8c97f4b42b6ba1c26dfc728dbc8

DarkDimius commented:
can we add those quirks as test cases?

lgo commented:
Yep! It may either require a new assertion type to the test runner or only some minor changes about the exhaustiveness of the assertion `^^^` markers, depending on how closely we want the results will match the current `def` + `usage` assertions and results.

Both find-definition and find-references also have these quirks, where they highlight the whole `def` line or referenced call. Notably, the typescript LSP does not seem to do this.

jvilk-stripe commented on test/helpers/position_assertions.cc:
```diff
@@ -546,6 +571,128 @@ void UsageAssertion::check(const UnorderedMap<string, shared_ptr<core::File>> &s
+    }
+
+    if (assertResponseMessage(id, *responses.at(0))) {
+        auto &respMsg = responses.at(0)->asResponse();
```
Could we move the shared logic into a helper method or maybe a shared parent class b/w UsageAssertion and HighlightAssertion? The basic premise is identical:

* Make query at a specific position.
* Check that the resulting locations match a set of expected locations.

I know that `DocumentHighlight` returns highlights that are not `Location`s, but you could convert them into `Location`s trivially and define a common interface. That'll also prevent the need for a separate `getLine`.

jvilk-stripe commented on test/helpers/position_assertions.cc:
```diff
@@ -546,6 +571,128 @@ void UsageAssertion::check(const UnorderedMap<string, shared_ptr<core::File>> &s
+    auto locSourceLine = getLine(sourceFileContents, uriPrefix, queryLoc);
+    string locFilename = uriToFilePath(uriPrefix, queryLoc.uri);
+
+    auto referenceParams =
```
This variable is unused.

jvilk-stripe commented on main/lsp/requests/document_highlight.cc:
```diff
@@ -0,0 +1,100 @@
+            auto resp = move(queryResponses[0]);
+            // N.B.: Ignores literals.
+            // If file is untyped, only supports find reference requests from constants and class definitions.
+            if (auto constResp = resp->isConstant()) {
```
I wonder if there's a clean way to share this logic between documentHighlight and references. I can't think of any atm that doesn't involve a nasty `variant`.

jvilk-stripe commented:
>  It may either require a new assertion type to the test runner or only some minor changes about the exhaustiveness of the assertion ^^^ markers, depending on how closely we want the results will match the current def + usage assertions and results.

I'm +1 to the def + usage assertion markers needing to match the exact range that Sorbet returns. We can make that change in a separate PR if it seems like a lot for this one. Note that I'll be picky about how we surface that new type of test error; it's important to me that test failures are easy-to-understand. :)

The quirks that you found are outside the scope of this MVP PR, but are something we should address for all features.

DarkDimius commented:
Closing it as stale, please feel free to reopen.
Would love this to get over the finish line, but it's not on the critical path for the team currently.

lgo commented:
@jvilk-stripe I return with a rebased PR! I also addressed the earlier outstanding items. Thanks for the refactors in `lsp`, it's certaintly a lot easier to work in :). I think you'll need to re-open the PR to have it update from the fork.

I did some digging into the `Location`s used around here. As I understand it, we use the `declLoc` from these that come from the parsing step. Specifically, the problem cases from before are that we just want location of the `SymbolRef` for these -- and it primarily happened on `Declaration` nodes and `Arg`-type `Reference` nodes. I imagine we'd need to carry through a `symbolLoc` or know how to lazily compute it.

Is that a correct understanding? I'm just digging to understand what needs to be done to make this LSP feature usable. As you mentioned, let's get this in w/o that fix yet.
